### ISSUE TITLE
Varnish crashes with separate VCL files and POST requests

### ISSUE THREAD

**Issue by ernestojpg** [varnishcache/varnish-cache#2367]
Hello!

I'm getting this segmentation fault in Varnish 5.1.2 when receiving POST requests in a simple configuration with two VCL files:

```
Error: Child (96827) died signal=6
Error: Child (96827) Panic at: Mon, 10 Jul 2017 21:38:34 GMT
Assert error in child_sigsegv_handler(), mgt/mgt_child.c line 271:
  Condition(Segmentation fault by instruction at 0x0) not true.
version = varnish-5.1.2 revision 6ece695, vrt api = 6.0
ident = Darwin,16.6.0,x86_64,-jnone,-smalloc,-smalloc,-hcritbit,kqueue
now = 3282252.606878 (mono), 1499722714.474813 (real)
Backtrace:
  0x102ca1a61: 0   varnishd                            0x0000000102ca1a61 pan_ic + 437
  0x102cd08f3: 0   varnishd                            0x0000000102cd08f3 child_sigsegv_handler + 102
  0x7fff8b1cfb3a: 0   libsystem_platform.dylib            0x00007fff8b1cfb3a _sigtramp + 26
  0x102d4f4c8: 0   varnishd                            0x0000000102d4f4c8 waiter + 5544
  0x102ca4b83: 0   varnishd                            0x0000000102ca4b83 vrb_pull + 199
  0x102ca51a7: 0   varnishd                            0x0000000102ca51a7 VRB_Ignore + 53
  0x102ca62da: 0   varnishd                            0x0000000102ca62da CNT_Request + 3754
  0x102cc504c: 0   varnishd                            0x0000000102cc504c http1_req + 491
  0x102cbd532: 0   varnishd                            0x0000000102cbd532 WRK_Thread + 1655
  0x102cbce5d: 0   varnishd                            0x0000000102cbce5d pool_thread + 51
thread = (cache-worker)
thr.req = 0x7f8401025420 {
  vxid = 2, transport = HTTP/1 {
    state = HTTP1::Proc
  }
  step = R_STP_FETCH,
  req_body = R_BODY_TAKEN,
  restarts = 0, esi_level = 0,
  sp = 0x7f8400e01480 {
    fd = 16, vxid = 1,
    t_open = 1499722714.474208,
    t_idle = 1499722714.474208,
    transport = HTTP/1 {
      state = HTTP1::Proc
    }
    client = 127.0.0.1 49537,
    privs = 0x7f8400e014e8 {
    },
  },
  worker = 0x700006a63c50 {
    stack = {0x700006a64000 -> 0x700006a58000},
    ws = 0x700006a63e50 {
      id = \"wrk\",
      {s, f, r, e} = {0x700006a633e0, +0, 0x0, +2040},
    },
    VCL::method = PASS,
    VCL::return = fetch,
    VCL::methods = {RECV, PASS, HASH},
  },
  ws = 0x7f8401025608 {
    id = \"req\",
    {s, f, r, e} = {0x7f8401027408, +160, 0x0, +57328},
  },
  http_conn = 0x7f8401025530 {
    fd = 16 (@0x7f8400e01498),
    doclose = NULL,
    ws = 0x7f8401025608 {
      [Already dumped, see above]
    },
    {rxbuf_b, rxbuf_e} = {0x7f8401027408, 0x7f84010274a0},
    {pipeline_b, pipeline_e} = {0x7f84010274a0, 0x7f84010274a2},
    content_length = 2,
    body_status = length,
    first_byte_timeout = 0.000000,
    between_bytes_timeout = 0.000000,
  },
  http[req] = 0x7f84010256a0 {
    ws = 0x7f8401025608 {
      [Already dumped, see above]
    },
    hdrs {
      \"POST\",
      \"/test/\",
      \"HTTP/1.1\",
      \"Host: localhost:8080\",
      \"User-Agent: curl/7.51.0\",
      \"Accept: */*\",
      \"Content-Length: 2\",
      \"Content-Type: application/x-www-form-urlencoded\",
    },
  },
  vcl = {
    name = \"vcl_test_alt\",
    busy = 1,
    discard = 0,
    state = auto,
    temp = warm,
    conf = {
      srcname = {
        \"/usr/local/etc/varnish/test_alt.vcl\",
        \"Builtin\",
      },
    },
  },
  vmods = {
    std = {Varnish 5.1.2 6ece695, 6.0},
  },
  objcore[BODY] = 0x7f8400d02a60 {
    refcnt = 1,
    flags = {busy, private},
    exp_flags = {},
    boc = 0x7f8400d029a0 {
      refcnt = 1,
      state = invalid,
      vary = 0x0,
      stevedore_priv = 0x0,
    },
    exp = {0.000000, 0.000000, 0.000000, 0.000000},
    objhead = 0x7f8400c04190,
    stevedore = 0x7f8400f009c0 (malloc Transient) {
      Simple = 0x7f8400d02b00,
      Obj = 0x7f8400d02b88 {priv=0x7f8400d02b80, ptr=0x7f8400d02b00, len=120, space=128},
      LEN = 0x0...0,
      VXID = 0x00000000,
      FLAGS = 0x00,
      GZIPBITS = 0x0...0,
      LASTMODIFIED = 0x0...0,
      VARY = {len=0, ptr=0x0},
      HEADERS = {len=0, ptr=0x0},
    },
  },
  objcore[REQ] = 0x7f8400d02900 {
    refcnt = 1,
    flags = {pass, private},
    exp_flags = {},
    exp = {1499722714.474562, 0.000000, 0.000000, 0.000000},
    objhead = 0x7f8400c04190,
    stevedore = 0x7f8400f009c0 (malloc Transient) {
      Simple = 0x7f8400f02fe0,
      Obj = 0x7f8400f02ba8 {priv=0x7f8400f02ba0, ptr=0x7f8400f02fe0, len=272, space=272},
      LEN = 0x0000000000000116,
      VXID = 0x00000003,
      FLAGS = 0x00,
      GZIPBITS = 0x0...0,
      LASTMODIFIED = 0x41d658fcf6800000,
      VARY = {len=0, ptr=0x0},
      HEADERS = {len=0, ptr=0x7f8400f03058},
      Body = 0x7f8400f02ce8 {priv=0x7f8400f02ce0, ptr=0x7f8400f030f0, len=278, space=278},
    },
  },
  flags = {
  },
},
thr.busyobj = 0x0 {
},
```

## Steps to Reproduce

Create two separate VCL files:

**test_main.vcl**
```
vcl 4.0;

backend default_backend {
    .host = "127.0.0.1";
    .port = "80";
}

sub vcl_recv {

    if (req.url ~ "^/test/") {
        return (vcl(label_test_alt));
    }

}
```

**test_alt.vcl**
```
vcl 4.0;

import std;

backend default_backend {
    .host = "127.0.0.1";
    .port = "80";
}

sub vcl_recv {

    std.log("We are here!! URL is now: " + req.url);

}
```

Then we need a `clifile` like this:

**test_clifile**
```
vcl.load vcl_test_alt /usr/local/etc/varnish/test_alt.vcl
vcl.label label_test_alt vcl_test_alt

vcl.load vcl_test_main /usr/local/etc/varnish/test_main.vcl
vcl.use vcl_test_main
```

Now, just start Varnish using the clifile:

`/usr/local/sbin/varnishd -n /usr/local/var/varnish -f "" -s malloc,1G -T 127.0.0.1:2000 -a 0.0.0.0:8080 -F -I /usr/local/etc/varnish/test_clifile`

Finally, let's make a curl POST request to the local Varnish:

`curl -d '{}' 'http://localhost:8080/test/'`

Note that we are sending a body in the request. We should see the segmentation fault in Varnish.

## Environment

* Version used: Varnish stable 5.1.2 (bottled by Homebrew)
* Operating System: macOS Sierra 10.12.5
* Homebrew 1.2.3

It is also reproducible in **Varnish 5.1.1** and CentOS Linux release 7.2.1511

**Comment by michbsd** [comment]
Looks like the same problem I'm having #2361

**Comment by fgsch** [comment]
@ernestojpg as with #2361, can you check if reverting 3e015a1 fixes the issue?

**Comment by ernestojpg** [comment]
@fgsch , that commit was part of Varnish 5.1.2, but I also was able to reproduce the issue in Varnish 5.1.1. But I will double check.

**Comment by fgsch** [comment]
@ernestojpg any updates?

**Comment by ernestojpg** [comment]
Hi @fgsch !

I have reproduced the issue in the following versions of Varnish:

- Varnish 5.1.0 (bottled by Homebrew), macOS Sierra 10.12.5
- Varnish 5.1.1 (bottled by Homebrew), macOS Sierra 10.12.5, and CentOS Linux release 7.2.1511
- Varnish 5.1.2 (bottled by Homebrew), macOS Sierra 10.12.5
- Varnish 5.1.3 (bottled by Homebrew), macOS Sierra 10.12.5

So it seems that the problem was not introduced with [3e015a1](https://github.com/varnishcache/varnish-cache/commit/3e015a1a8e256c65d4536390b6db098740ca9a5e).

Thank you very much for taking a look to this.

**Comment by michbsd** [comment]
I concur. I had to downgrade to "varnish-5.0.0 revision 99d036f" to get it working

**Comment by fgsch** [comment]
@ernestojpg @michbsd thanks for checking. This helps narrowing it down.

**Comment by daghf** [comment]
Here's a vtc based on the above description.

```
varnishtest "POST and return(vcl(..))"

server s1 {
	rxreq
	txresp
} -start

varnish v1 -vcl+backend {} -start

shell "echo 'vcl 4.0; backend foo { .host = \"${s1_addr}\"; .port = \"${s1_port}\"; }' > ${tmpdir}/_r02367_alt.vcl"
shell "echo 'vcl 4.0; backend foo { .host = \"${s1_addr}\"; .port = \"${s1_port}\"; } sub vcl_recv { return (vcl(label_alt)); }' > ${tmpdir}/_r02367_main.vcl"

varnish v1 -cliok "vcl.load vcl_alt ${tmpdir}/_r02367_alt.vcl"
varnish v1 -cliok "vcl.label label_alt vcl_alt"
varnish v1 -cliok "vcl.load vcl_main ${tmpdir}/_r02367_main.vcl"
varnish v1 -cliok "vcl.use vcl_main"

client c1 {
	txreq -req POST -body "asdf"
	rxresp
	expect resp.status == 200
} -run
```

**Comment by daghf** [comment]
PR with a proposed fix in #2385 

Now also with a less hacky testcase (thanks @Dridi)
